nixos/modules/virtualisation/proxmox-image.nix

```
@@ -0,0 +1,165 @@
+        type = types.commas;
+        default = "virtio=00:00:00:00:00:00,bridge=vmbr0,firewall=1";
+        description = ''
+          Configuration for the default interface. When restoring from VMA, check the
+          "unique" box to ensure device mac is randomized.
+        '';
+      };
+      serial0 = mkOption {
+        type = types.enum [ "socket" ];
+        default = "socket";
+        description = ''
+          Enable serial terminal
+        '';
+      };
+      agent = mkOption {
+        type = types.bool;
+        apply = x: if x then "1" else "0";
+        default = true;
+        description = ''
+          Expect guest to have qemu agent running
+        '';
+      };
+    };
+    qemuExtraConf = mkOption {
+      type = with types; attrsOf (oneOf [ str int ]);
+      default = {};
+      example = literalExpression ''{
+        cpu = "host";
+        onboot = 1;
+      }'';
+      description = ''
+        Additional options appended to qemu-server.conf
+      '';
+    };
+    filenameSuffix = mkOption {
+      type = types.str;
+      default = config.proxmox.qemuConf.name;
+      example = "999-nixos_template";
+      description = ''
+        Filename of the image will be vzdump-qemu-''${filenameSuffix}.vma.zstd.
+        This will also determine the default name of the VM on restoring the VMA.
+        Start this value with a number if you want the VMA to be detected as a backup of
+        any specific VMID.
+      '';
+    };
+  };
+
+  config = let
+    cfg = config.proxmox;
+    cfgLine = name: value: ''
+      ${name}: ${builtins.toString value}
+    '';
+    cfgFile = fileName: properties: pkgs.writeTextDir fileName ''
+      # generated by NixOS
+      ${lib.concatStrings (lib.mapAttrsToList cfgLine properties)}
+      #qmdump#map:virtio0:drive-virtio0:local-lvm:raw:
+    '';
+  in {
+    system.build.VMA = import ../../lib/make-disk-image.nix {
+      name = "proxmox-${cfg.filenameSuffix}";
+      postVM = let
+        # Build qemu with PVE's patch that adds support for the VMA format
+        vma = pkgs.qemu_kvm.overrideAttrs ( super: {
+          patches = let
+            rev = "cc707c362ea5c8d832aac270d1ffa7ac66a8908f";
+            path = "debian/patches/pve/0025-PVE-Backup-add-vma-backup-format-code.patch";
+            vma-patch = pkgs.fetchpatch {
+              url = "https://git.proxmox.com/?p=pve-qemu.git;a=blob_plain;hb=${rev};f=${path}";
+              sha256 = "1z467xnmfmry3pjy7p34psd5xdil9x0apnbvfz8qbj0bf9fgc8zf";
+            };
+          in super.patches ++ [ vma-patch ];
+          buildInputs = super.buildInputs ++ [ pkgs.libuuid ];
+        });
+      in
+      ''
+        ${vma}/bin/vma create vzdump-qemu-${cfg.filenameSuffix}.vma -c ${cfgFile "qemu-server.conf" (cfg.qemuConf // cfg.qemuExtraConf)}/qemu-server.conf drive-virtio0=$diskImage
+        rm $diskImage
+        ${pkgs.zstd}/bin/zstd vzdump-qemu-${cfg.filenameSuffix}.vma
+        mv vzdump-qemu-${cfg.filenameSuffix}.vma.zst $out/
```

```suggestion
        ${vma}/bin/vma create "vzdump-qemu-${cfg.filenameSuffix}.vma" \
          -c "${cfgFile "qemu-server.conf" (cfg.qemuConf // cfg.qemuExtraConf)}/qemu-server.conf" \
          "drive-virtio0=$diskImage"
        rm "$diskImage"
        ${pkgs.zstd}/bin/zstd "vzdump-qemu-${cfg.filenameSuffix}.vma"
        mv "vzdump-qemu-${cfg.filenameSuffix}.vma.zst" "$out/"
```

I always suggest to quote variables because scripts break in unforseeable ways when something expand into a string with spaces.

---

Quoting qemu-server.conf is unnecessary. The filename is always "qemu-server.conf", and it is in a nix store path.

---

I don't want to force this change, but quoting is always needed IMO, unless you explicitely want word splitting. I know, it's a terrible default but it's the way shells work.

Btw, the Nix store path is usually something like `/nix/store`, but it's not a strict requirement: for example , people in HPC commonly have it under their home directory, which may contain spaces.

---

People int he HPC community use nix from their home directories containing spaces? That is impressive given
1. [Running nix with non-standard store paths](https://github.com/NixOS/nix/issues/1971) is an open issue, and the best hack available involves rebuilding everything 
and more importantly,
2. Bootstrap tools, one of the most basic requirement for running nix builds, [will fail if your store path has spaces](https://asciinema.org/a/mjGIwgbtU52KYaxQjEDGgSWjj) because of this line: https://github.com/NixOS/nixpkgs/blob/354ff094e42e16055d4b2a3624c2928d7fc02bd4/pkgs/stdenv/linux/bootstrap-tools/scripts/unpack-bootstrap-tools.sh#L3